WCAG-pagina 4.1.1 Parsen (#899)
nl-design-system/documentatie: docs/wcag/4.1.1.mdx

---
title: WCAG-succescriterium 4.1.1 Parsen
hide_title: true
hide_table_of_contents: false
sidebar_label: 4.1.1 Parsen
pagination_label: WCAG-succescriterium 4.1.1 Parsen
description: Zorg dat er geen fouten in de HTML-code zijn.
slug: 4.1.1
keywords:
  - WCAG
---

{/* @license CC0-1.0 */}

import CTAGebruikersonderzoek from "./_cta_gebruikersonderzoek.md";
import WCAGFooterInfo from "./_wcag_footer_info.md";

# WCAG-succescriterium 4.1.1 Parsen

> **Let op**: Dit succescriterium is misschien in jouw situatie niet meer van toepassing. Het is in WCAG-versie 2.2 verwijderd. In versie 2.1 is een [opmerking](https://www.w3.org/TR/WCAG21/#h-note-26) toegevoegd dat het criterium automatisch voldoet wanneer HTML of XML wordt gebruikt, maar die is niet ‘normatief’.

Reviewer: ```suggestion
**Let op**: De werkgroep die WCAG maakt heeft 4.1.1 de afgelopen jaren besproken en vastgesteld dat dit criterium tegenwoordig niet meer relevant is, omdat browsers en hulptechnologieën intussen zijn verbeterd (zie de WCAG FAQ](https://www.w3.org/WAI/standards-guidelines/wcag/faq/#parsing411)). Het criterium misschien in jouw situatie niet meer van toepassing. In WCAG-versie 2.2 is het helemaal verwijderd. In versie 2.1 is een [opmerking](https://www.w3.org/TR/WCAG21/#h-note-26) toegevoegd dat het criterium automatisch voldoet wanneer HTML of XML wordt gebruikt.
```

slug: 4.1.1
keywords:
  - WCAG
---

{/* @license CC0-1.0 */}

import CTAGebruikersonderzoek from "./_cta_gebruikersonderzoek.md";
import WCAGFooterInfo from "./_wcag_footer_info.md";

# WCAG-succescriterium 4.1.1 Parsen

**Let op**: Dit succescriterium is in WCAG-versie 2.2 vervallen. Maar we vinden het belangrijk hier te vermelden en uit te leggen, omdat het niet goed toepassen grote consequences heeft voor de werking van formuleren.

## W3C referenties WCAG-versie 2.1

- Engelse tekst van het WCAG-succescriterium: [<span lang="en">4.1.1 Parsing</span>](https://www.w3.org/TR/WCAG21/#parsing).
- Nederlandse vertaling van het WCAG-succescriterium: [4.1.1 Parsen](https://www.w3.org/Translations/WCAG21-nl/#parsen).
- Engelstalige informatie op <span lang="en">How to Meet WCAG</span>: [<span lang="en">Quick Reference 4.1.1 Parsing</span>](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.1#parsing).
- Engelstalige toelichting: [<span lang="en">Understanding SC 4.1.1 Parsing</span>](https://www.w3.org/WAI/WCAG21/Understanding/parsing.html).

## Uitleg

Wat is parsen? Officieel: het analyseren van informatie aan de hand van losse componenten, of het analyseren van de syntax.

> Voor WCAG betekent het dat de HTML-code moet kloppen, foutloos moet zijn. Dan werkt een webpagina beter. Vooral voor formulieren is het belangrijk dat ID's uniek zijn op een webpagina. Er zijn 4 criteria gedefinieerd.

Author: Uit het SC>
Ik heb het SC in vieren gesplits, want het zijn 4 verschillende onderwerpen:

> In content die geïmplementeerd is met opmaaktalen hebben elementen volledige begin- en eindtags, zijn elementen genest volgens hun specificatie, bevatten elementen geen dubbele attributen en zijn alle ID's uniek, behalve waar de specificatie deze eigenschappen toelaat. 

Dus 
1: hebben elementen volledige begin- en eindtags
2: zijn elementen genest volgens hun specificatie
3: bevatten elementen geen dubbele attributen
4: zijn alle ID's uniek
Want dat zijn de 4 creteria waaop getekst wordt.

### 1. Elementen zijn volgens hun specificaties afgesloten

Denk hier bijvoorbeeld aan paragrafen, headings, divs, spans en andere elementen die een sluittag vereisen.

> Een `<p>` moet afsluiten met een `</p>`.

Author: ik heb er een div van gemaakt

---
title: WCAG-succescriterium 4.1.1 Parsen
hide_title: true
hide_table_of_contents: false
sidebar_label: 4.1.1 Parsen
pagination_label: WCAG-succescriterium 4.1.1 Parsen
description: Zorg dat er geen fouten in de HTML-code zijn.
slug: 4.1.1
keywords:
  - WCAG
---

{/* @license CC0-1.0 */}

import CTAGebruikersonderzoek from "./_cta_gebruikersonderzoek.md";
import WCAGFooterInfo from "./_wcag_footer_info.md";

# WCAG-succescriterium 4.1.1 Parsen

> **Let op**: Dit succescriterium is in WCAG-versie 2.2 verwijderd. Pas wanneer de opvolger van [EN_301_549 3.2.1](https://nl.wikipedia.org/wiki/EN_301_549) overstapt van WCAG 2.1 naar 2.2 dan is dit succescriterium niet meer wettelijk verplicht.

Reviewer: hier zou ik ook de opmerking in WCAG 2.1 willen noemen en dan misschien toch niet het wettelijke stukje, om het overzichtelijk te houden, aannemende dat men zelf al nadenkt over welke versie ze nastreven?

```suggestion
**Let op**: Dit succescriterium is misschien in jouw situatie niet meer van toepassing. Het is in WCAG-versie 2.2 verwijderd. In versie 2.1 is een [opmerking](https://www.w3.org/TR/WCAG21/#h-note-26) toegevoegd dat het criterium automatisch voldoet wanneer HTML of XML wordt gebruikt.
```

Author: is aangepast

## W3C referenties WCAG-versie 2.1

- Engelse tekst van het WCAG-succescriterium: [<span lang="en">4.1.1 Parsing</span>](https://www.w3.org/TR/WCAG21/#parsing).
- Nederlandse vertaling van het WCAG-succescriterium: [4.1.1 Parsen](https://www.w3.org/Translations/WCAG21-nl/#parsen).
- Engelstalige informatie op <span lang="en">How to Meet WCAG</span>: [<span lang="en">Quick Reference 4.1.1 Parsing</span>](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.1#parsing).
- Engelstalige toelichting: [<span lang="en">Understanding SC 4.1.1 Parsing</span>](https://www.w3.org/WAI/WCAG21/Understanding/parsing.html).

## Uitleg

> Voor WCAG 2.1 moet de "markup parsen" geen problemen opleveren, dat betekent dat de HTML-code foutloos moet zijn. Dan werkt een webpagina beter.

Reviewer: ik vind deze niet zo lekker lezen, voorstel, en zou hier WCAG-versie weglaten. Ook gaat het niet alleen om HTML-code, maar ook om bv HTML-code of web content in andere opmaaktalen. En is er in WCAG 2.1 een note die zegt dat content in HTML  altijd voldoet.

```suggestion
De “markup” van pagina's, denk aan bijvoorbeeld HTML-code, moet voldoen aan bepaalde voorwaarden. Zo wordt de samenwerking tussen webpagina's, browsers en hulptechnologieën minder foutgevoelig.
```

Author: is aangepast

> Wat betekent parsen eigenlijk? Officieel: het analyseren van informatie aan de hand van losse componenten, of het analyseren van de syntax.

Reviewer: Deze zou ik weglaten, ik denk dat het niet per se duidelijker wordt, en het woord “officieel” suggereert teveel?

```suggestion
```

Author: is aangepast

Er zijn 4 voorwaarden gedefinieerd.

### 1. Elementen zijn volgens hun specificaties afgesloten

Denk hier bijvoorbeeld aan paragrafen, headings, divs, spans en andere elementen die een sluittag vereisen.
Een `<div>` moet afsluiten met een `</div>`.

### 2. Elementen zijn genest volgens hun specificaties

Verkeerd nesten van elementen kan voorkomen bij interactieve elementen: een button die een tweede button bevat.
Of bij het verkeerd combineren van elementen: een label die een link bevat.
Daarnaast moeten bijvoorbeeld lists en definition lists goed opgebouwd zijn.

Silvestar Bistrović beschrijft een aantal [<span lang="en">Common nesting issues in HTML</span>](https://www.htmhell.dev/adventcalendar/2022/20/).

### 3. Elementen bevatten geen dubbele attributen

> Combineer waardes in attributen en herhaal de attributen zelf niet.

Author: is aangepast

> Bijvoorbeeld voor het toekennen van CSS-classes:

Reviewer: ```suggestion
Vermijd bijvoorbeeld meerdere `class`-attributen:
```

Author: is aangepast

```html
<!-- Foute code, niet gebruiken -->
<h1 class="heading" class="x-large">Contact</h1>
```

> Voeg de classes samen, gescheiden door een spatie:

Reviewer: ``suggestion
In het geval van het `class`-attribuut scheid je meerdere waarden met een spatie:
```

Author: is aangepast

```html
<h1 class="heading x-large">Contact</h1>
```

### 4. Alle ID's zijn uniek, behalve waar de specificatie deze eigenschappen toelaat

Dit is voor formulieren belangrijk. Labels en formuliervelden zijn via een ID aan elkaar gekoppeld. Komt een ID meerdere keren voor op een pagina, dan kan het verkeerde label bij een formulierveld worden voorgelezen.

In code:

```html
<label for="kleur">Wat is je lievelingskleur?</label> <input type="text" id="kleur" name="lievelingskleur" />
```

Bij de richtlijnen over labels staat hoe deze [for/id relatie](/richtlijnen/formulieren/labels/de-forid-relatie) werkt.

Een ander probleem is het toevoegen van twee keer hetzelfde formulier met hetzelfde formulier-ID, de foutafhandeling staat dan misschien bij het verkeerde formulier.

## Gebruikersonderzoek

<CTAGebruikersonderzoek />

## Hoe te testen

- Controleer de webpagina op de [Markup Validation Service](https://validator.w3.org/).

> - Controleer de webpagina met de [<span lang="en">WCAG 2.1 parsing error bookmarklet</span>](https://www.tpgi.com/wcag-2-0-parsing-error-bookmarklet/) van Steve Faulkner

Reviewer: zou ik weglaten omdat er nu een note in 2.1 zit die zegt dat parsing ook niet meer getest hoeft te worden voor HTML

```suggestion
```

Author: is aangepast

- [axe Devtools](https://www.deque.com/axe/browser-extensions/) markeert enkele (maar niet alle) parse-fouten in de HTML.

Voor developers:

- Gebruik een [HTML-linter](https://github.com/caramelomartins/awesome-linters?tab=readme-ov-file#html), of een linter naar keuze die ook voor correcte HTML test.
- Bij twijfel: check de '<span lang="en">Permitted content</span>' bij de '<span lang="en">Technical Summary</span>' van een HTML-element op de MDN-website. Bijvoorbeeld [de technische samenvatting voor een link](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#technical_summary).

## Veelgemaakte fouten

> ### Fout: meerdere keren hetzelfde Gravity Forms formulier op een webpagina.

Reviewer: Zou deze vendor-neutral opschrijven:

```suggestion
### Fout: dubbele ID's doordat een  formulier meerdere keren wordt gebruikt.
```

Author: is aangepast

De WordPress formulieren-plugin [Gravity Forms](https://docs.gravityforms.com/accessibility-checklist-for-gravity-forms/) geeft elk formulier een eigen ID. Voeg je hetzelfde formulier twee keer toe op een pagina, dan worden de fouten afgehandeld bij het bovenste formulier. Dit is verwarrend als je net het onderste formulier aan het invullen bent.

> Een formulier dat vaak twee keer wordt toegevoegd, is een aanmeldformulier voor de nieuwsbrief. Wil je toch hetzelfde formulier meerdere keren op een pagina, maak dan verschillende formulieren hiervoor aan.

Reviewer: “hiervoor” kan weg?

```suggestion
Een formulier dat vaak twee keer wordt toegevoegd, is een aanmeldformulier voor de nieuwsbrief. Wil je toch hetzelfde formulier meerdere keren op een pagina, maak dan verschillende formulieren aan.
```

Author: is aangepast

pagination_label: WCAG-succescriterium 4.1.1 Parsen
description: Zorg dat er geen fouten in de HTML-code zijn.
slug: 4.1.1
keywords:
  - WCAG
---

{/* @license CC0-1.0 */}

import CTAGebruikersonderzoek from "./_cta_gebruikersonderzoek.md";
import WCAGFooterInfo from "./_wcag_footer_info.md";

# WCAG-succescriterium 4.1.1 Parsen

**Let op**: Dit succescriterium is in WCAG-versie 2.2 vervallen. Maar we vinden het belangrijk hier te vermelden en uit te leggen, omdat het niet goed toepassen grote consequences heeft voor de werking van formuleren.

## W3C referenties WCAG-versie 2.1

- Engelse tekst van het WCAG-succescriterium: [<span lang="en">4.1.1 Parsing</span>](https://www.w3.org/TR/WCAG21/#parsing).
- Nederlandse vertaling van het WCAG-succescriterium: [4.1.1 Parsen](https://www.w3.org/Translations/WCAG21-nl/#parsen).
- Engelstalige informatie op <span lang="en">How to Meet WCAG</span>: [<span lang="en">Quick Reference 4.1.1 Parsing</span>](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.1#parsing).
- Engelstalige toelichting: [<span lang="en">Understanding SC 4.1.1 Parsing</span>](https://www.w3.org/WAI/WCAG21/Understanding/parsing.html).

## Uitleg

> Wat is parsen? Officieel: het analyseren van informatie aan de hand van losse componenten, of het analyseren van de syntax.

Reviewer 2: Dit is wel een heel technische binnenkomer 😅 Ik denk dat we de volgende alinea ietsjes kunnen tweaken, zodat deze niet nodig is.

Author: Ik vond het wel leuk om parsing uit te leggen, want dat was het eerste wat ik me afvroeg toen ik deze pgina ging schrijven Wat betekent parsen eigenlijk.

Author: ik draai het wel om.

> Voor WCAG betekent het dat de HTML-code moet kloppen, foutloos moet zijn. Dan werkt een webpagina beter. Vooral voor formulieren is het belangrijk dat ID's uniek zijn op een webpagina. Er zijn 4 criteria gedefinieerd.

Reviewer 2: ```suggestion
Voor WCAG 2.1 moet de "markup parsen" geen problemen opleveren, dat betekent dat de HTML-code foutloos moet zijn. Dan werkt een webpagina beter. Vooral voor formulieren is het belangrijk dat ID's uniek zijn op een webpagina. Er zijn 4 criteria gedefinieerd.
```

### 1. Elementen zijn volgens hun specificaties afgesloten

Denk hier bijvoorbeeld aan paragrafen, headings, divs, spans en andere elementen die een sluittag vereisen.
Een `<p>` moet afsluiten met een `</p>`.

### 2. Elementen zijn genest volgens hun specificaties

Verkeerd nesten van elementen kan voorkomen bij interactieve elementen: een button die een tweede button bevat.
Of bij het verkeerd combineren van elementen: een label die een link bevat.
Daarnaast moeten bijvoorbeeld lists en definition lists goed opgebouwd zijn.

Silvestar Bistrović beschrijft een aantal [<span lang="en">Common nesting issues in HTML</span>](https://www.htmhell.dev/adventcalendar/2022/20/).

### 3. Elementen bevatten geen dubbele attributen

Combineer waardes in attributen en herhaal de attributen zelf niet.
Bijvoorbeeld voor het toekennen van CSS-classes:

```html
<!-- Foute code, niet gebruiken -->
<h1 class="heading" class="x-large">Contact</h1>
```

Voeg de classes samen, gescheiden door een spatie:

```html
<h1 class="heading x-large">Contact</h1>
```

> 

Author: Ik zou PHP niet de schuld willen geven van dubbele ID's :-)
En om iedereen naar React te verwijzen gaat mij een brug te ver.
Ik wil het graag zo algemeen mogelijk houden.

> 

Reviewer: Ja eens met Rian; het interessante context en goed punt re string templates Robbert, maar voert denk ik te ver voor de richtlijn.
